crates/data_components/src/spark_connect/federation.rs

```diff
@@ -0,0 +1,121 @@
+use futures::Stream;
+use spark_connect_rs::errors::SparkError;
+
+use super::SparkConnectTableProvider;
+
+impl SparkConnectTableProvider {
+    fn create_federated_table_source(
+        self: Arc<Self>,
+    ) -> DataFusionResult<Arc<dyn FederatedTableSource>> {
+        let table_name = self.table_reference.to_string();
+        tracing::trace!(
+            table_name,
+            %self.table_reference,
+            "create_federated_table_source"
+        );
+        let schema = Arc::clone(&self.schema);
+        let fed_provider = Arc::new(SQLFederationProvider::new(self));
+        Ok(Arc::new(SQLTableSource::new_with_schema(
+            fed_provider,
+            table_name,
+            schema,
+        )?))
+    }
+
+    pub fn create_federated_table_provider(
+        self: Arc<Self>,
+    ) -> DataFusionResult<FederatedTableProviderAdaptor> {
+        let table_source = Self::create_federated_table_source(Arc::clone(&self))?;
+        Ok(FederatedTableProviderAdaptor::new_with_provider(
+            table_source,
+            self,
+        ))
+    }
+}
+
+#[async_trait]
+impl SQLExecutor for SparkConnectTableProvider {
+    fn name(&self) -> &str {
+        "SparkConnect"
+    }
+
+    fn compute_context(&self) -> Option<String> {
+        Some(self.join_push_down_context.clone())
+    }
+
+    fn dialect(&self) -> Arc<dyn Dialect> {
+        Arc::new(
+            CustomDialectBuilder::new()
+                .with_interval_style(datafusion::sql::unparser::dialect::IntervalStyle::SQLStandard)
+                .with_identifier_quote_style('`')
+                .build(),
+        )
+    }
+
+    fn execute(
+        &self,
+        query: &str,
+        schema: SchemaRef,
+    ) -> DataFusionResult<SendableRecordBatchStream> {
+        Ok(Box::pin(RecordBatchStreamAdapter::new(
+            schema,
+            spark_query_to_stream(self.dataframe.clone().sparkSession(), query.to_string()),
+        )))
+    }
+
+    async fn table_names(&self) -> DataFusionResult<Vec<String>> {
+        Err(DataFusionError::NotImplemented(
+            "table inference not implemented".to_string(),
+        ))
+    }
+
+    async fn get_table_schema(&self, table_name: &str) -> DataFusionResult<SchemaRef> {
+        Ok(self
+            .dataframe
+            .clone()
+            .sparkSession()
+            .table(table_name)
+            .map_err(map_error_to_datafusion_err)?
+            .limit(0)
+            .collect()
+            .await
+            .map_err(map_error_to_datafusion_err)?
+            .schema())
+    }
+}
+
+#[allow(clippy::needless_pass_by_value)]
+fn spark_query_to_stream(
+    session: Arc<spark_connect_rs::SparkSession>,
+    query: String,
+) -> impl Stream<Item = DataFusionResult<RecordBatch>> {
+    let session = Arc::clone(&session);
+
+    stream! {
+        let data = session
+            .sql(&query)
+            .await
+            .map_err(map_error_to_datafusion_err)?
+            .collect()
```

Comment:
This uses the existing way to collect data which is not streaming yet, will address this in future.